File: app/spiffsconfig.cpp
#include <SmingCore/SmingCore.h>
#include "spiffsconfig.h"

SpiffsConfigStorage NetConfig;

const String NET_SETTINGS_FILE = "net.conf";
const String WIFISSID_SETTINGS_FILES[MAX_WIFI_SETS] = {"wifi0.ssid","wifi1.ssid","wifi2.ssid"};
const String WIFIPASS_SETTINGS_FILES[MAX_WIFI_SETS] = {"wifi0.pass","wifi1.pass","wifi2.pass"};
const String MQTTCLIENT_SETTINGS_FILE = "mqtt.client";
const String MQTTLVL1TOPIC_SETTINGS_FILE = "mqttlvl1.conf";
const String MQTTUSER_SETTINGS_FILE = "mqtt.user";
const String MQTTPASS_SETTINGS_FILE = "mqtt.pass";
const String MQTTBROKER_SETTINGS_FILE = "mqttbrkr.conf";
const String AUTHTOKEN_SETTINGS_FILE = "authtoken";
const String DOMAINNAME_SETTINGS_FILE = "domname.conf";
const String USEDHCP_SETTINGS_FILE[MAX_WIFI_SETS] = {"dhcp0.flag","dhcp1.flag","dhcp2.flag"};
const String DNS_SERVERS_FILE = "dns";
const String LOCATION_SETTINGS_FILE = "loc.conf";


void SpiffsConfigStorage::load()
{
	if (exist())
	{
		uint32_t netsettings[4];
		file_t f = fileOpen(NET_SETTINGS_FILE, eFO_ReadOnly);
		fileRead(f, (void*) netsettings, 4*sizeof(uint32_t));
		fileClose(f);
		ip = IPAddress(netsettings[0]);
		netmask = IPAddress(netsettings[1]);
		gw = IPAddress(netsettings[2]);
		mqtt_port = (uint16_t)(netsettings[3]);
		if (fileExist(DNS_SERVERS_FILE))
		{
			uint32_t dnssettings[DNS_MAX_SERVERS];
			f = fileOpen(DNS_SERVERS_FILE, eFO_ReadOnly);
			fileRead(f, (void*) dnssettings, DNS_MAX_SERVERS*sizeof(uint32_t));
			fileClose(f);
			for (uint32_t d=0; d<DNS_MAX_SERVERS; d++)
				dns[d] = IPAddress(dnssettings[d]);
		}
		for (uint32_t wifi_settings_num=0; wifi_settings_num<MAX_WIFI_SETS; wifi_settings_num++)
		{
			wifi_ssid[wifi_settings_num] = fileGetContent(WIFISSID_SETTINGS_FILES[wifi_settings_num]);
			wifi_pass[wifi_settings_num] = fileGetContent(WIFIPASS_SETTINGS_FILES[wifi_settings_num]);
			enabledhcp[wifi_settings_num] = fileExist(USEDHCP_SETTINGS_FILE[wifi_settings_num]);
		}
		mqtt_broker = fileGetContent(MQTTBROKER_SETTINGS_FILE);
		mqtt_clientid = fileGetContent(MQTTCLIENT_SETTINGS_FILE);
		mqtt_user = fileGetContent(MQTTUSER_SETTINGS_FILE);
		mqtt_pass = fileGetContent(MQTTPASS_SETTINGS_FILE);
		authtoken = fileGetContent(AUTHTOKEN_SETTINGS_FILE);
		domainname = fileGetContent(DOMAINNAME_SETTINGS_FILE);
		mqtt_lvl1_topic = fileGetContent(MQTTLVL1TOPIC_SETTINGS_FILE);
		
		location = fileGetContent(LOCATION_SETTINGS_FILE);

	}
}

void SpiffsConfigStorage::save()
{
	uint32_t netsettings[4] = {ip,netmask,gw, (uint32_t)mqtt_port};
	file_t f = fileOpen(NET_SETTINGS_FILE, eFO_WriteOnly | eFO_CreateNewAlways);
	fileWrite(f, (void*) netsettings, 4*sizeof(uint32_t));
	fileClose(f);
	uint32_t dnssettings[DNS_MAX_SERVERS];
	for (uint32_t d=0; d<DNS_MAX_SERVERS; d++)
		dnssettings[d]=(uint32_t)dns[d];
	f = fileOpen(DNS_SERVERS_FILE, eFO_WriteOnly | eFO_CreateNewAlways);
	fileWrite(f, (void*) dnssettings, DNS_MAX_SERVERS*sizeof(uint32_t));
	fileClose(f);
	for (uint32_t ws=0; ws<MAX_WIFI_SETS; ws++)
	{
		if (wifi_ssid[ws].length() > 0) {
			fileSetContent(WIFISSID_SETTINGS_FILES[ws], wifi_ssid[ws]);
			fileSetContent(WIFIPASS_SETTINGS_FILES[ws], wifi_pass[ws]);
		}
		if (enabledhcp[ws])
			fileSetContent(USEDHCP_SETTINGS_FILE[ws], "true");
		else
			fileDelete(USEDHCP_SETTINGS_FILE[ws]);
	}
	fileSetContent(MQTTBROKER_SETTINGS_FILE, mqtt_broker);
	fileSetContent(MQTTCLIENT_SETTINGS_FILE, mqtt_clientid);
	fileSetContent(MQTTUSER_SETTINGS_FILE, mqtt_user);
	fileSetContent(MQTTPASS_SETTINGS_FILE, mqtt_pass);
	fileSetContent(AUTHTOKEN_SETTINGS_FILE, authtoken);
	fileSetContent(MQTTLVL1TOPIC_SETTINGS_FILE, mqtt_lvl1_topic);
	fileSetContent(DOMAINNAME_SETTINGS_FILE, domainname);

	fileSetContent(LOCATION_SETTINGS_FILE, location);
}

bool SpiffsConfigStorage::exist() { return fileExist(NET_SETTINGS_FILE); }

File: spiffsconfig.py
#!/usr/bin/python3

from ipaddress import ip_address
import os
import struct

def writeConfig(ip,nm,gw,wifi0_ssid,wifi0_pass,mqtt_broker,mqtt_level1topic,mqtt_clientid,mqtt_user,mqtt_pass,authtoken,dhcp0=True,dhcp1=True,dhcp2=True,domainname="lan",mqtt_port=1883,wifi1_ssid=b"",wifi1_pass=b"",wifi2_ssid=b"",wifi2_pass=b"",dns=[],location=""):
    NET_SETTINGS_FILE = "net.conf"
    WIFISSID0_SETTINGS_FILE = "wifi0.ssid"
    WIFIPASS0_SETTINGS_FILE = "wifi0.pass"
    WIFISSID1_SETTINGS_FILE = "wifi1.ssid"
    WIFIPASS1_SETTINGS_FILE = "wifi1.pass"
    WIFISSID2_SETTINGS_FILE = "wifi2.ssid"
    WIFIPASS2_SETTINGS_FILE = "wifi2.pass"
    DOMAINNAME_SETTINGS_FILE = "domname.conf"
    MQTTCLIENT_SETTINGS_FILE = "mqtt.client"
    MQTTUSER_SETTINGS_FILE = "mqtt.user"
    MQTTPASS_SETTINGS_FILE = "mqtt.pass"
    MQTTBROKER_SETTINGS_FILE = "mqttbrkr.conf"
    MQTTLVL1TOPIC_SETTINGS_FILE = "mqttlvl1.conf"
    AUTHTOKEN_SETTINGS_FILE = "authtoken"
    USEDHCP0_SETTINGS_FILE = "dhcp0.flag"
    USEDHCP1_SETTINGS_FILE = "dhcp1.flag"
    USEDHCP2_SETTINGS_FILE = "dhcp2.flag"
    DNS_SERVERS_FILE = "dns"
    DNS_MAX_SERVERS=2
    LOCATION_SETTINGS_FILE = "loc.conf"
    DIR="./files/"
    if not os.path.isdir(DIR):
        os.mkdir(DIR)
    with open(os.path.join(DIR, NET_SETTINGS_FILE),"wb") as fh:
        fh.write(struct.pack(">III", int(ip_address(ip)), int(ip_address(nm)), int(ip_address(gw))))
        fh.write(struct.pack("<I",  int(mqtt_port)))
    if dns:
        with open(os.path.join(DIR, DNS_SERVERS_FILE),"wb") as fh:
            for ipidx in range(0,DNS_MAX_SERVERS):
                try:
                    fh.write(struct.pack(">I", int(ip_address(dns[ipidx]))))
                except IndexError:
                    fh.write(struct.pack(">I", int(ip_address("8.8.4.4"))))
    else:
        try:
            os.unlink(os.path.join(DIR, DNS_SERVERS_FILE))
        except:
            pass
    with open(os.path.join(DIR, WIFISSID0_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi0_ssid)
    with open(os.path.join(DIR, WIFIPASS0_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi0_pass)
    with open(os.path.join(DIR, WIFISSID1_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi1_ssid)
    with open(os.path.join(DIR, WIFIPASS1_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi1_pass)
    with open(os.path.join(DIR, WIFISSID2_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi2_ssid)
    with open(os.path.join(DIR, WIFIPASS2_SETTINGS_FILE),"wb") as fh:
        fh.write(wifi2_pass)
    with open(os.path.join(DIR, DOMAINNAME_SETTINGS_FILE),"wb") as fh:
        fh.write(domainname)
    with open(os.path.join(DIR, MQTTBROKER_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_broker)
    with open(os.path.join(DIR, MQTTCLIENT_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_clientid)
    with open(os.path.join(DIR, MQTTUSER_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_user)
    with open(os.path.join(DIR, MQTTPASS_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_pass)
    with open(os.path.join(DIR, AUTHTOKEN_SETTINGS_FILE),"wb") as fh:
        fh.write(authtoken)
    with open(os.path.join(DIR, MQTTLVL1TOPIC_SETTINGS_FILE),"wb") as fh:
        fh.write(mqtt_level1topic)
    if dhcp0:
        with open(os.path.join(DIR, USEDHCP0_SETTINGS_FILE),"wb") as fh:
            fh.write(b"true")
    else:
        try:
            os.unlink(os.path.join(DIR, USEDHCP0_SETTINGS_FILE))
        except:
            pass
    if dhcp1:
        with open(os.path.join(DIR, USEDHCP1_SETTINGS_FILE),"wb") as fh:
            fh.write(b"true")
    else:
        try:
            os.unlink(os.path.join(DIR, USEDHCP1_SETTINGS_FILE))
        except:
            pass
    if dhcp2:
        with open(os.path.join(DIR, USEDHCP2_SETTINGS_FILE),"wb") as fh:
            fh.write(b"true")
    else:
        try:
            os.unlink(os.path.join(DIR, USEDHCP2_SETTINGS_FILE))
        except:
            pass
    with open(os.path.join(DIR, LOCATION_SETTINGS_FILE),"wb") as fh:
        fh.write(location)


### Usage ###
##
## Create a .py File with content below and run
##
## Example:
# from spiffsconfig import writeConfig
# writeConfig(
#     ip="192.168.1.11",
#     nm="255.255.255.0",
#     gw="192.168.1.1",
#     dns=["1.1.1.1","8.8.8.8"],
#     domainname=b"localdomain",
#     dhcp0=True,
#     wifi0_ssid=b"MyWifi",
#     wifi0_pass=b"MyWifiPassword",
#     location=b"livingroom",
#     mqtt_port=1883,
#     mqtt_broker=b"mqtt.localdomain",
#     mqtt_clientid=b"smokedetector1",
#     mqtt_level1topic=b"sensors",
#     mqtt_user=b"",
#     mqtt_pass=b"",
#     authtoken=b"none")
